// starrail.js/dist/models/light_cone/LightCone.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const LightConeData_1 = __importDefault(require("./LightConeData"));
/**
 * @en
 */
class LightCone {
    /**
     * @param data
     * @param client
     */
    constructor(data, client) {
        this.client = client;
        this._data = data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.lightConeData = new LightConeData_1.default(json.getAsNumber("tid"), this.client);
        this.level = json.getAsNumber("level");
        this.ascension = json.getAsNumberWithDefault(0, "promotion");
        this.superimposition = this.lightConeData.superimpositions[json.getAsNumber("rank") - 1];
        this.basicStats = this.lightConeData.getStatsByLevel(this.ascension, this.level);
        this.extraStats = this.lightConeData.getSuperimpositionStats(this.superimposition.level);
    }
}
exports.default = LightCone;
