// starrail.js/dist/models/enka/StarRailCharacterBuild.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const enka_system_1 = require("enka-system");
const Character_1 = __importDefault(require("../character/Character"));
/**
 * @extends {CharacterBuild}
 */
class StarRailCharacterBuild extends enka_system_1.CharacterBuild {
    /**
     * @param data
     * @param client
     * @param username
     * @param hash
     */
    constructor(data, client, username, hash) {
        super();
        this._data = data;
        this.client = client;
        this.enkaUserInfo = { username: username, hash: hash };
        const json = new config_file_js_1.JsonReader(this._data);
        this.id = json.getAsNumber("id");
        this.name = json.getAsString("name");
        this.order = json.getAsNumber("order");
        this.isLive = json.getAsBoolean("live");
        this.isPublic = json.getAsBoolean("public");
        this.character = new Character_1.default(json.getAsJsonObject("avatar_data"), client);
        this.imageUrl = json.getAsNullableString("image");
        this.hoyoType = json.getAsNumber("hoyo_type");
        this.url = `https://enka.network/u/${this.enkaUserInfo.username}/${this.enkaUserInfo.hash}/${this.character.characterData.id}/${this.id}`;
    }
}
exports.default = StarRailCharacterBuild;
