// starrail.js/dist/models/relic/RelicMainStatGroup.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const AssetsNotFoundError_1 = __importDefault(require("../../errors/AssetsNotFoundError"));
const StatProperty_1 = __importDefault(require("../StatProperty"));
/**
 * @en RelicMainStatGroup
 */
class RelicMainStatGroup {
    /**
     * @param id
     * @param client
     */
    constructor(id, client) {
        this.id = id;
        this.client = client;
        const _data = client.cachedAssetsManager.getStarRailCacheData("RelicMainAffixConfig")[this.id];
        if (!_data)
            throw new AssetsNotFoundError_1.default("RelicMainStatGroup", this.id);
        this._data = _data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.mainStats = json.mapObject((_, v) => {
            return {
                id: v.getAsNumber("AffixID"),
                groupId: this.id,
                statProperty: new StatProperty_1.default(v.getAsString("Property"), this.client),
                baseValue: v.getAsNumber("BaseValue", "Value"),
                levelValue: v.getAsNumber("LevelAdd", "Value"),
            };
        });
    }
}
exports.default = RelicMainStatGroup;
