// starrail.js/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.otherStatPropertyTypes = exports.statPropertyTypes = exports.combatTypeIconColors = exports.nonNullable = exports.getStableHash = exports.fetchJSON = exports.isStatPropertyType = exports.sumStats = exports.UserIcon = exports.StarRailUser = exports.StatPropertyValue = exports.StatProperty = exports.Path = exports.CombatType = exports.RelicType = exports.RelicSubStatGroup = exports.RelicSetBonus = exports.RelicSet = exports.RelicMainStatGroup = exports.RelicExpType = exports.RelicData = exports.RelicSubStat = exports.RelicMainStat = exports.Relic = exports.LightConeSuperimposition = exports.LightConeExpType = exports.LightConeData = exports.LightCone = exports.StarRailCharacterBuild = exports.Eidolon = exports.OverallStatList = exports.StatList = exports.CharacterStats = exports.CharacterData = exports.Character = exports.LeveledSkillTreeNode = exports.SkillTreeNode = exports.SkillLevel = exports.LeveledSkill = exports.Skill = exports.TextAssets = exports.ImageAssets = exports.DynamicTextAssets = exports.AssetsNotFoundError = exports.StarRail = exports.ObjectKeysManager = exports.CachedAssetsManager = void 0;
const CachedAssetsManager_1 = __importDefault(require("./client/CachedAssetsManager"));
exports.CachedAssetsManager = CachedAssetsManager_1.default;
const ObjectKeysManager_1 = __importDefault(require("./client/ObjectKeysManager"));
exports.ObjectKeysManager = ObjectKeysManager_1.default;
const StarRail_1 = __importDefault(require("./client/StarRail"));
exports.StarRail = StarRail_1.default;
const AssetsNotFoundError_1 = __importDefault(require("./errors/AssetsNotFoundError"));
exports.AssetsNotFoundError = AssetsNotFoundError_1.default;
const DynamicTextAssets_1 = __importDefault(require("./models/assets/DynamicTextAssets"));
exports.DynamicTextAssets = DynamicTextAssets_1.default;
const ImageAssets_1 = __importDefault(require("./models/assets/ImageAssets"));
exports.ImageAssets = ImageAssets_1.default;
const TextAssets_1 = __importDefault(require("./models/assets/TextAssets"));
exports.TextAssets = TextAssets_1.default;
const Skill_1 = __importStar(require("./models/character/skill/Skill"));
exports.Skill = Skill_1.default;
Object.defineProperty(exports, "LeveledSkill", { enumerable: true, get: function () { return Skill_1.LeveledSkill; } });
const SkillLevel_1 = __importDefault(require("./models/character/skill/SkillLevel"));
exports.SkillLevel = SkillLevel_1.default;
const SkillTreeNode_1 = __importStar(require("./models/character/skill/SkillTreeNode"));
exports.SkillTreeNode = SkillTreeNode_1.default;
Object.defineProperty(exports, "LeveledSkillTreeNode", { enumerable: true, get: function () { return SkillTreeNode_1.LeveledSkillTreeNode; } });
const Character_1 = __importDefault(require("./models/character/Character"));
exports.Character = Character_1.default;
const CharacterData_1 = __importDefault(require("./models/character/CharacterData"));
exports.CharacterData = CharacterData_1.default;
const CharacterStats_1 = __importStar(require("./models/character/CharacterStats"));
exports.CharacterStats = CharacterStats_1.default;
Object.defineProperty(exports, "StatList", { enumerable: true, get: function () { return CharacterStats_1.StatList; } });
Object.defineProperty(exports, "OverallStatList", { enumerable: true, get: function () { return CharacterStats_1.OverallStatList; } });
const Eidolon_1 = __importDefault(require("./models/character/Eidolon"));
exports.Eidolon = Eidolon_1.default;
const StarRailCharacterBuild_1 = __importDefault(require("./models/enka/StarRailCharacterBuild"));
exports.StarRailCharacterBuild = StarRailCharacterBuild_1.default;
const LightCone_1 = __importDefault(require("./models/light_cone/LightCone"));
exports.LightCone = LightCone_1.default;
const LightConeData_1 = __importDefault(require("./models/light_cone/LightConeData"));
exports.LightConeData = LightConeData_1.default;
const LightConeExpType_1 = __importDefault(require("./models/light_cone/LightConeExpType"));
exports.LightConeExpType = LightConeExpType_1.default;
const LightConeSuperimposition_1 = __importDefault(require("./models/light_cone/LightConeSuperimposition"));
exports.LightConeSuperimposition = LightConeSuperimposition_1.default;
const Relic_1 = __importStar(require("./models/relic/Relic"));
exports.Relic = Relic_1.default;
Object.defineProperty(exports, "RelicMainStat", { enumerable: true, get: function () { return Relic_1.RelicMainStat; } });
Object.defineProperty(exports, "RelicSubStat", { enumerable: true, get: function () { return Relic_1.RelicSubStat; } });
const RelicData_1 = __importDefault(require("./models/relic/RelicData"));
exports.RelicData = RelicData_1.default;
const RelicExpType_1 = __importDefault(require("./models/relic/RelicExpType"));
exports.RelicExpType = RelicExpType_1.default;
const RelicMainStatGroup_1 = __importDefault(require("./models/relic/RelicMainStatGroup"));
exports.RelicMainStatGroup = RelicMainStatGroup_1.default;
const RelicSet_1 = __importDefault(require("./models/relic/RelicSet"));
exports.RelicSet = RelicSet_1.default;
const RelicSetBonus_1 = __importDefault(require("./models/relic/RelicSetBonus"));
exports.RelicSetBonus = RelicSetBonus_1.default;
const RelicSubStatGroup_1 = __importDefault(require("./models/relic/RelicSubStatGroup"));
exports.RelicSubStatGroup = RelicSubStatGroup_1.default;
const RelicType_1 = __importDefault(require("./models/relic/RelicType"));
exports.RelicType = RelicType_1.default;
const CombatType_1 = __importDefault(require("./models/CombatType"));
exports.CombatType = CombatType_1.default;
const Path_1 = __importDefault(require("./models/Path"));
exports.Path = Path_1.default;
const StatProperty_1 = __importStar(require("./models/StatProperty"));
exports.StatProperty = StatProperty_1.default;
Object.defineProperty(exports, "StatPropertyValue", { enumerable: true, get: function () { return StatProperty_1.StatPropertyValue; } });
const StarRailUser_1 = __importDefault(require("./models/StarRailUser"));
exports.StarRailUser = StarRailUser_1.default;
const UserIcon_1 = __importDefault(require("./models/UserIcon"));
exports.UserIcon = UserIcon_1.default;
// functions
var CharacterStats_2 = require("./models/character/CharacterStats");
Object.defineProperty(exports, "sumStats", { enumerable: true, get: function () { return CharacterStats_2.sumStats; } });
var StatProperty_2 = require("./models/StatProperty");
Object.defineProperty(exports, "isStatPropertyType", { enumerable: true, get: function () { return StatProperty_2.isStatPropertyType; } });
var axios_utils_1 = require("./utils/axios_utils");
Object.defineProperty(exports, "fetchJSON", { enumerable: true, get: function () { return axios_utils_1.fetchJSON; } });
var hash_utils_1 = require("./utils/hash_utils");
Object.defineProperty(exports, "getStableHash", { enumerable: true, get: function () { return hash_utils_1.getStableHash; } });
var ts_utils_1 = require("./utils/ts_utils");
Object.defineProperty(exports, "nonNullable", { enumerable: true, get: function () { return ts_utils_1.nonNullable; } });
// constants
var CombatType_2 = require("./models/CombatType");
Object.defineProperty(exports, "combatTypeIconColors", { enumerable: true, get: function () { return CombatType_2.combatTypeIconColors; } });
var StatProperty_3 = require("./models/StatProperty");
Object.defineProperty(exports, "statPropertyTypes", { enumerable: true, get: function () { return StatProperty_3.statPropertyTypes; } });
Object.defineProperty(exports, "otherStatPropertyTypes", { enumerable: true, get: function () { return StatProperty_3.otherStatPropertyTypes; } });
