// starrail.js/dist/models/Path.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const TextAssets_1 = __importDefault(require("./assets/TextAssets"));
const AssetsNotFoundError_1 = __importDefault(require("../errors/AssetsNotFoundError"));
const ImageAssets_1 = __importDefault(require("./assets/ImageAssets"));
/**
 * @en Path
 */
class Path {
    /**
     * @param id
     * @param client
     */
    constructor(id, client) {
        this.id = id;
        this.client = client;
        const _data = client.cachedAssetsManager.getStarRailCacheData("AvatarBaseType")[this.id];
        if (!_data)
            throw new AssetsNotFoundError_1.default("Path", this.id);
        this._data = _data;
        const json = new config_file_js_1.JsonReader(this._data);
        this.name = new TextAssets_1.default(json.getAsNumber("BaseTypeText", "Hash"), this.client);
        this.description = new TextAssets_1.default(json.getAsNumber("BaseTypeDesc", "Hash"), this.client);
        this.icon = new ImageAssets_1.default(json.getAsString("BaseTypeIcon"), this.client);
    }
}
exports.default = Path;
