// starrail.js/dist/models/StarRailUser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.platformMap = void 0;
const config_file_js_1 = require("config_file.js");
const enka_system_1 = require("enka-system");
const Character_1 = __importDefault(require("./character/Character"));
const UserIcon_1 = __importDefault(require("./UserIcon"));
exports.platformMap = {
    1: "IOS",
    2: "ANDROID",
    3: "PC",
    11: "PS5",
};
/** @extends {User} */
class StarRailUser extends enka_system_1.User {
    /**
     * @param data
     * @param client
     */
    constructor(data, client) {
        var _a;
        const json = new config_file_js_1.JsonReader(data);
        super(json);
        this.client = client;
        const detailInfo = json.get("detailInfo");
        this.uid = (_a = detailInfo.getAsNumberWithDefault(null, "uid")) !== null && _a !== void 0 ? _a : Number(json.getValue("uid"));
        this.nickname = detailInfo.getAsString("nickname");
        this.signature = detailInfo.getAsStringWithDefault(null, "signature");
        this.icon = new UserIcon_1.default(detailInfo.getAsNumber("headIcon"), this.client);
        this.level = detailInfo.getAsNumber("level");
        this.equilibriumLevel = detailInfo.getAsNumberWithDefault(0, "worldLevel");
        const platform = detailInfo.getValue("platform");
        this.platform = typeof platform === "number" ? exports.platformMap[platform] : typeof platform === "string" ? platform : null;
        this.friends = detailInfo.getAsNumberWithDefault(0, "friendCount");
        const recordInfo = detailInfo.get("recordInfo");
        this.achievements = recordInfo.getAsNumberWithDefault(0, "achievementCount");
        this.characterCount = recordInfo.getAsNumber("avatarCount");
        this.lightConeCount = recordInfo.getAsNumberWithDefault(0, "equipmentCount");
        const challengeInfo = recordInfo.get("challengeInfo");
        this.forgottenHall = {
            memory: challengeInfo.getAsNumberWithDefault(0, "scheduleMaxLevel"),
            memoryOfChaos: challengeInfo.getAsNumberWithDefault(0, "noneScheduleMaxLevel"),
            memoryOfChaosId: challengeInfo.getAsNumberWithDefault(null, "scheduleGroupId"),
        };
        this.simulatedUniverse = recordInfo.getAsNumberWithDefault(0, "maxRogueChallengeScore");
        const avatarDetailList = detailInfo.getAsJsonArrayWithDefault([], "avatarDetailList");
        this.starfaringCompanions = avatarDetailList.filter(c => !c["_assist"]).map(c => new Character_1.default(c, this.client));
        this.supportCharacters = (() => {
            // mihomo
            if (detailInfo.has("assistAvatarList"))
                return detailInfo.getAsJsonArray("assistAvatarList").map(c => new Character_1.default(c, this.client));
            // enka
            return avatarDetailList.filter(c => c["_assist"]).map(c => new Character_1.default(c, this.client));
        })();
        this.enkaUserHash = json.getAsStringWithDefault(null, "owner", "hash");
    }
    /**  */
    getCharacters() {
        const characters = [...this.supportCharacters, ...this.starfaringCompanions];
        return Array.from(new Set(characters));
    }
}
exports.default = StarRailUser;
