// starrail.js/dist/models/assets/DynamicTextAssets.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const config_file_js_1 = require("config_file.js");
const TextAssets_1 = __importDefault(require("./TextAssets"));
const defaultDynamicData = {
    paramList: [],
};
/**
 * @en DynamicTextAssets
 * TextAssets which has placeholders in the text.
 */
class DynamicTextAssets extends TextAssets_1.default {
    // readonly _dynamicTest: string | null;
    /**
     * @param id
     * @param data
     * @param client
     */
    constructor(id, data, client) {
        super(id, client);
        this.dynamicData = (0, config_file_js_1.bindOptions)(defaultDynamicData, data);
        // this._dynamicTest = this.getNullableReplacedText();
    }
    /**
     * @param replaceWith
     * @param lang
     * @throws AssetsNotFoundError
     */
    getReplacedData(replaceWith = [], lang) {
        function isEnabled(key) {
            return replaceWith.length === 0 || replaceWith.includes(key);
        }
        const usedParamIndices = [];
        let text = this.get(lang);
        if (isEnabled("paramList")) {
            text = text.replace(/<unbreak>#(\d+)\[(.+?)\](%?)<\/unbreak>/g, (_, paramIndexText, valueType, percent) => {
                const paramIndex = parseInt(paramIndexText) - 1;
                const isPercent = percent === "%";
                const value = this.dynamicData.paramList[paramIndex] * (isPercent ? 100 : 1);
                usedParamIndices.push(paramIndex);
                const fix = valueType === "i" ? 0
                    : /^f\d+$/.test(valueType) ? Number(valueType.replace("f", ""))
                        : null;
                if (fix === null) {
                    // TODO: remove this
                    console.error(`Unknown valueType ${valueType} in DynamicTextAssets.`);
                    throw new Error(`Unknown valueType ${valueType} in DynamicTextAssets.`);
                }
                const fixedValue = value.toFixed(fix);
                return fixedValue + percent;
            });
        }
        return { text, usedParamIndices };
    }
    /**
     * @param replaceWith
     * @param lang
     * @returns null instead of throwing AssetsNotFoundError.
     */
    getNullableReplacedData(replaceWith = [], lang) {
        try {
            return this.getReplacedData(replaceWith, lang);
        }
        catch (e) {
            return null;
        }
    }
    /**
     * @param replaceWith
     * @param lang
     * @throws AssetsNotFoundError
     */
    getReplacedText(replaceWith = [], lang) {
        return this.getReplacedData(replaceWith, lang).text;
    }
    /**
     * @param replaceWith
     * @param lang
     * @returns null instead of throwing AssetsNotFoundError.
     */
    getNullableReplacedText(replaceWith = [], lang) {
        try {
            return this.getReplacedText(replaceWith, lang);
        }
        catch (e) {
            return null;
        }
    }
}
exports.default = DynamicTextAssets;
