// starrail.js/dist/models/assets/ImageAssets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @en ImageAssets
 */
class ImageAssets {
    /**
     * @param path
     * @param client
     */
    constructor(path, client) {
        var _a;
        this.client = client;
        this.path = path;
        this.imageBaseUrl = (_a = [...client.options.imageBaseUrls].sort((a, b) => b.priority - a.priority).find(url => url.regexList.some(regex => regex.test(this.path)))) !== null && _a !== void 0 ? _a : null;
        this.url = (() => {
            if (this.path === "" || this.imageBaseUrl == null)
                return "";
            const convertedPath = convertPathForImageBaseUrl(this.imageBaseUrl, this.path);
            if ("customParser" in this.imageBaseUrl)
                return `${this.imageBaseUrl.url}/${this.imageBaseUrl.customParser(convertedPath)}`;
            return `${this.imageBaseUrl.url}/${convertedPath}`;
        })();
        this.isAvailable = this.url !== null && this.url !== undefined && this.url !== "";
    }
}
exports.default = ImageAssets;
function convertPathForImageBaseUrl(imageBaseUrl, path) {
    const split = path.split("/");
    switch (imageBaseUrl.filePath) {
        case "UPPER_CAMEL_CASE":
            return path;
        case "LOWER_CASE":
            return split.slice(0, -1).join("/").toLowerCase() + "/" + split.slice(-1)[0];
        case "NONE":
            return split.slice(-1)[0];
    }
}
